### PR TITLE
SR-1: added all tasks for homework_1

### PR DESCRIPTION
**Pull request by Sofiiar**
**Description:**
- added `tests` directory in `homework_1`,   `tests` directory includes:

1. test_not_reliable_function.py .
2. test_not_reliable_function_2.py .
3. test_not_reliable_function_3.py .
4. test_use_config.py.
5. test_check_password.py

- source code for tests is placed into `homework_1` directory

**Screenshot:s**
Screenshot 2024-04-20 at 21 08 05

### CODE REVIEW

File: tests/sradu/homework_1/tests/test_check_password.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    assert check_password("PSSWRD1!") is False
+
+def test_check_password_failure_no_number():
+    assert check_password("Psswrdd!") is False
+
+def test_check_password_failure_no_symbol():
+    assert check_password("Psswrdd1") is False
+
+def test_chekc_password_failure_too_many_characters():
+    assert check_password("Psswrd111111111111111111111111!") is False
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_check_password.py:24]: use parametrize to group tests

File: tests/sradu/homework_1/tests/test_not_reliable_function.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def test_not_reliable_success(mocker):
+    mocker.patch('tests.sradu.homework_1.not_reliable_function.random.random', return_value=0.6)
+    assert not_reliable() == 0.6
+
+
+def test_not_reliable_failure(mocker):
+    mocker.patch('tests.sradu.homework_1.not_reliable_function.random.random', return_value=0.4)
+    with pytest.raises(RuntimeError) as exception_info:
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_not_reliable_function.py:12]: line 14 redundant

with pytest.raises(RuntimeError, match="Res is less than 0.5:  0.4"):
    ...

File: tests/sradu/homework_1/tests/test_not_reliable_function.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+from tests.sradu.homework_1.not_reliable_function import not_reliable
+
+
+def test_not_reliable_success(mocker):
+    mocker.patch('tests.sradu.homework_1.not_reliable_function.random.random', return_value=0.6)
+    assert not_reliable() == 0.6
+
+
+def test_not_reliable_failure(mocker):
+    mocker.patch('tests.sradu.homework_1.not_reliable_function.random.random', return_value=0.4)
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_not_reliable_function.py:11]: mocker.patch('random.random', return_value=0.4)


or 


import random
mocker.patch.object(random, 'random', return_value=0.4)

File: tests/sradu/homework_1/tests/test_not_reliable_function_2.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import pytest
+from tests.sradu.homework_1.not_reliable_function_2 import not_reliable
+
+
+def test_not_reliable_success(mocker):
+    mocker.patch('tests.sradu.homework_1.not_reliable_function_2.random', return_value=0.5)
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_not_reliable_function_2.py:6]: you can patch like this: 


from tests.sradu.homework_1 import not_reliable_function_2 as module
mocker.patch.obejct(module, 'random', return_value=0.5)

File: tests/sradu/homework_1/tests/test_use_config.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def test_is_debug_on_returns_original_value():
+    assert if_debug_on() is True
+
+def test_if_debug_on_returns_false(mocker):
+    mocker.patch.dict(CONFIG, {'debug': False})
+    assert  if_debug_on() is False
+
+def test_if_debug_on_raises_key_error(mocker):
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_use_config.py:19]: 2 empty lienes between functions in module

File: tests/sradu/homework_1/tests/test_use_config.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import pytest
+from tests.sradu.homework_1.use_config import if_debug_on, CONFIG
+
+
+@pytest.fixture(scope="function")
+def restore_config():
+    original_config = CONFIG.copy()
+    yield
+    CONFIG.update(original_config)
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_use_config.py:9]: no need in this fixture you use appropriate tools for patching
pytest handles revert for you 


try fixture.  monkeypatch
https://docs.pytest.org/en/latest/how-to/monkeypatch.html

File: tests/sradu/homework_1/tests/test_not_reliable_function_3.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    assert  tries == 1
+
+def test_not_reliable_success_maximum_tries(mocker):
+    sequence_of_returns = [0.1, 0.887, 0.0, 0.01, 0.1, 0.887, 0.0, 0.01, 0.01, 0.888]
+    mocker.patch('tests.sradu.homework_1.not_reliable_function_3.random', side_effect=sequence_of_returns)
+    result, tries = not_reliable()
+    assert result >= 0.888
+    assert tries == 10
+
+def test_not_reliable_failure(mocker):
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_not_reliable_function_3.py:18]: 2 empty lines between functions

**Review by kudinov-fedor — APPROVED** [review]

File: tests/sradu/homework_1/tests/test_not_reliable_function_3.py (hunk 7 above)
Comment (Sofiiar) [reply tests/sradu/homework_1/tests/test_not_reliable_function_3.py:20]: fixed

**Review by kudinov-fedor — APPROVED** [review]

File: tests/sradu/homework_1/tests/test_not_reliable_function.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def test_not_reliable_success(mocker):
+    mocker.patch.object(random, 'random', return_value=0.6)
+    assert not_reliable() == 0.6
+
+
+def test_not_reliable_failure(mocker):
+    mocker.patch.object(random, 'random', return_value=0.4)
+    with pytest.raises(RuntimeError, match=r"Res is less than 0.5:\s+0.4"):
```
Comment (kudinov-fedor) [inline tests/sradu/homework_1/tests/test_not_reliable_function.py:13]: why to use regex symbols if you can avoid them

r"Res is less than 0.5:  0.4"

**Pull request merged**